### PR TITLE
Feat: support to sync the revision and the workflow status

### PR DESCRIPTION
**Pull request by barnettZQG**
Signed-off-by: barnettZQG <barnett.zqg@gmail.com>

### Description of your changes

1. Sync the application revision to the datastore, this is useful to back up the application configuration. But currently, only sync the application spec.
2. Sync the workflow record to the datastore, users could resume the workflow via UI.
3. Do not allow to change the inner labels via API. such as `ux.oam.dev/from-namespace` `app.oam.dev/source-of-truth` `ux.oam.dev/synced-generation`
4. Optimized the logic during application synchronization.

Fixes #

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: pkg/apiserver/domain/service/workflow.go [hunk 1]
```diff
@@ -358,45 +358,59 @@ func (w *workflowServiceImpl) SyncWorkflowRecord(ctx context.Context) error {
...
-		if err := w.KubeClient.Get(ctx, types.NamespacedName{
-			Name:      fmt.Sprintf("record-%s-%s", appName, record.Name),
-			Namespace: record.Namespace,
-		}, cr); err != nil {
-			klog.ErrorS(err, "failed to get controller revision", "oam app name", appName, "workflow name", record.WorkflowName, "record name", record.Name)
-			continue
+		var revision = &model.ApplicationRevision{AppPrimaryKey: record.AppPrimaryKey, Version: record.RevisionPrimaryKey}
+		if err := w.Store.Get(ctx, revision); err != nil {
+			if errors.Is(err, datastore.ErrRecordNotExist) {
+				return bcode.ErrApplicationRevisionNotExist
```
Comment (FogDong) [inline pkg/apiserver/domain/service/workflow.go:386]: Shouldn't we set record to terminated if the application revision is not found?

File: pkg/apiserver/domain/service/workflow.go [hunk 2]
```diff
@@ -358,45 +358,59 @@ func (w *workflowServiceImpl) SyncWorkflowRecord(ctx context.Context) error {
...
+			return err
 		}
-		appInRevision, err := util.RawExtension2Application(cr.Data)
-		if err != nil {
-			klog.ErrorS(err, "failed to get app data in controller revision", "controller revision name", cr.Name, "app name", appName, "workflow name", record.WorkflowName, "record name", record.Name)
-			continue
+
+		var appRevision v1beta1.ApplicationRevision
+		if err := w.KubeClient.Get(ctx, types.NamespacedName{Namespace: app.Namespace, Name: revision.RevisionCRName}, &appRevision); err != nil {
+			log.Logger.Warnf("failed to get the application revision %s", err.Error())
```
Comment (FogDong) [inline pkg/apiserver/domain/service/workflow.go:393]: Shouldn't we set record to terminated if the application revision is not found?

**Review by wonderflow — APPROVED** [review]

File: pkg/apiserver/domain/service/workflow.go (hunk 2 above)
Comment (barnettZQG) [reply pkg/apiserver/domain/service/workflow.go:397]: Good, already changed.

File: pkg/apiserver/domain/service/workflow.go [hunk 3]
```diff
@@ -358,45 +358,63 @@ func (w *workflowServiceImpl) SyncWorkflowRecord(ctx context.Context) error {
...
 		// try to sync the status from the controller revision
-		cr := &appsv1.ControllerRevision{}
-		if err := w.KubeClient.Get(ctx, types.NamespacedName{
-			Name:      fmt.Sprintf("record-%s-%s", appName, record.Name),
-			Namespace: record.Namespace,
-		}, cr); err != nil {
-			klog.ErrorS(err, "failed to get controller revision", "oam app name", appName, "workflow name", record.WorkflowName, "record name", record.Name)
+		var revision = &model.ApplicationRevision{AppPrimaryKey: record.AppPrimaryKey, Version: record.RevisionPrimaryKey}
+		if err := w.Store.Get(ctx, revision); err != nil {
+			log.Logger.Errorf("failed to get the application revision from database %s", err.Error())
```
Comment (FogDong) [inline pkg/apiserver/domain/service/workflow.go:385]: Same `setRecordTerminate` here?

Comment (barnettZQG) [reply pkg/apiserver/domain/service/workflow.go:401]: If the application revision is not found, set the record status to failure.

**Pull request merged**